Python/Product/PythonTools/PythonTools/Project/PythonProjectNode.cs

```diff
@@ -777,76 +780,104 @@ internal partial class PythonProjectNode :
+
+                // if we have no interpreter factories, and the active interpreter is the global default,
+                // add a node for it
+                if (InterpreterFactories.Count() == 0 && IsActiveInterpreterGlobalDefault && ActiveInterpreter.IsRunnable()) {
+                    var newNode = new InterpretersNode(
+                        this,
+                        ActiveInterpreter,
+                        isInterpreterReference: true,
+                        canDelete: false,
+                        isGlobalDefault: true
+                        );
+
+                    pythonEnvironmentsNode.AddChild(newNode);
                 }
-            }
 
-            foreach (var child in remaining) {
-                node.RemoveChild(child);
-            }
+                // add all the factories we have
+                foreach (var interpreterFactory in InterpreterFactories) {
+                    var isProjectSpecific = _vsProjectContext.IsProjectSpecific(interpreterFactory.Configuration);
+                    var canRemove = !this.IsAppxPackageableProject(); // Do not allow change python environment for UWP
+                    var canDelete = isProjectSpecific && Directory.Exists(interpreterFactory.Configuration.GetPrefixPath());
 
-            if (alwaysCollapse || ParentHierarchy == null) {
-                OnInvalidateItems(node);
-            } else {
-                bool wasExpanded = node.GetIsExpanded();
-                var expandAfter = node.AllChildren.Where(n => n.GetIsExpanded()).ToArray();
-                OnInvalidateItems(node);
-                if (wasExpanded) {
-                    node.ExpandItem(EXPANDFLAGS.EXPF_ExpandFolder);
+                    var newNode = new InterpretersNode(
+                        this,
+                        interpreterFactory,
+                        isInterpreterReference: !isProjectSpecific,
+                        canDelete,
+                        isGlobalDefault: false,
+                        canRemove
+                    );
+
+                    pythonEnvironmentsNode.AddChild(newNode);
+                }
+
+                // If the project is referencing interpreters that we can't find, add dummy nodes for them.
+                // This can include virtual environments that have been deleted, interpreters that have been uninstalled,
+                // or conda environments that are still being discovered asynchronously.
+                foreach (var id in InvalidInterpreterIds) {
+                    pythonEnvironmentsNode.AddChild(InterpretersNode.CreateAbsentInterpreterNode(this, id));
                 }
-                foreach (var child in expandAfter) {
-                    child.ExpandItem(EXPANDFLAGS.EXPF_ExpandFolder);
+
+                // Handle the expand/collapse state of the container node, plus each interpreter node
+                // TODO: Is this still needed? We cleared out all the previous nodes so the previous expanded info is gone?
```

Comment:
Yeah, I think this won't do anything useful now. If you want to preserve the expansion state on refresh it needs to be captured before clearing the nodes out.

Reply:
I decided to just expand the python environments node and keep all the child nodes collapsed.